### PR TITLE
feat(spark-core): handle locus URL redirects

### PR DESCRIPTION
**Pull request by saurjai3**
# Pull Request Template

## Description
Support Web Client Sharding (307 redirect)

Fixes # (issue)
https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-18307

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## Test Coverage
Verified all tests passed for spark-core and http-core packages locally.

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: packages/node_modules/@ciscospark/http-core/src/interceptors/http-status.js [hunk 1]
```diff
@@ -39,7 +39,7 @@ export default class HttpStatusInterceptor extends Interceptor {
    * @returns {Promise}
    */
   onResponse(options, response) {
-    if (response.statusCode && response.statusCode < 400) {
+    if (response.statusCode && (response.statusCode < 400 || (response.statusCode === 404 && response.options && response.options.qs && response.options.qs.alternateRedirect))) { // to handle locus redirects
```
Comment (bzang) [inline packages/node_modules/@ciscospark/http-core/src/interceptors/http-status.js]: style nitpick, can you break this into multiple lines to make it easier to parse at a glance?

File: packages/node_modules/@ciscospark/spark-core/src/interceptors/redirect.js [hunk 2]
```diff
@@ -60,6 +61,18 @@ export default class RedirectInterceptor extends Interceptor {
 
       return this.spark.request(options);
     }
+    else if (response.headers && response.body && response.body.errorCode === this.spark.config.LOCUS_REDIRECT_ERROR && response.body.location && response.options && response.options.qs && response.options.qs.alternateRedirect) {
```
Comment (bzang) [inline packages/node_modules/@ciscospark/spark-core/src/interceptors/redirect.js]: style nitpick, can you break this into multiple lines to make it easier to parse at a glance?

File: packages/node_modules/@ciscospark/spark-core/src/interceptors/redirect.js [hunk 3]
```diff
@@ -50,6 +50,7 @@ export default class RedirectInterceptor extends Interceptor {
    * @returns {Object}
    */
   onResponse(options, response) {
+    /* eslint-disable no-else-return */
```
Comment (bzang) [inline packages/node_modules/@ciscospark/spark-core/src/interceptors/redirect.js]: Is this eslint-disable necessary? Your changes don't seem to violate this rule.

File: packages/node_modules/@ciscospark/spark-core/src/config.js [hunk 4]
```diff
@@ -10,6 +10,8 @@ const IDBROKER_BASE_URL = process.env.IDBROKER_BASE_URL || 'https://idbroker.web
 
 export default {
   maxAppLevelRedirects: 10,
+  maxLocusRedirects: 5,
+  LOCUS_REDIRECT_ERROR: 2000002,
```
Comment (bzang) [inline packages/node_modules/@ciscospark/spark-core/src/config.js]: Did you mean to make this a user configurable option? Seems more like a constant that shouldn't be changeable by the developer.

File: packages/node_modules/@ciscospark/spark-core/src/interceptors/redirect.js (hunk 3 above)
Comment (saurjai3) [reply packages/node_modules/@ciscospark/spark-core/src/interceptors/redirect.js]: Yes, it gives an eslint error for no-else-return otherwise.

**Comment by saurjai3** [comment]
We are not using the internal-plugin-locus package inside the web client and as a result it is not imported as part of the recipe-private-web-client. I can write a new locus interceptor but still need to include it in recipe-private else the interceptor won't function. @mattnorris , any idea how to achieve that?

**Comment by mattnorris** [comment]
I think this is fine given the interceptor in question is specifically for redirects. @bzang I understand your comment about locus's role, but I think it could live either place. What do you think based on this new information?

**Comment by bzang** [comment]
@mattnorris @saurjai3 , I think that's reasonable! Thanks!

File: packages/node_modules/@ciscospark/spark-core/src/interceptors/redirect.js [hunk 5]
```diff
@@ -60,6 +62,26 @@ export default class RedirectInterceptor extends Interceptor {
...
+      && response.body
+      && response.body.errorCode === LOCUS_REDIRECT_ERROR
+      && response.body.location) {
+      options = clone(options);
+      if (response.options && response.options.qs) {
+        const newUrl = response.body.location.split('?');
+        options.uri = newUrl[0]; // params are already present in the qs
+      }
+      else {
+        options.uri = response.body.location;
```
Comment (saurjai3) [inline packages/node_modules/@ciscospark/spark-core/src/interceptors/redirect.js]: for GET requests

File: packages/node_modules/@ciscospark/spark-core/src/interceptors/redirect.js [hunk 6]
```diff
@@ -60,6 +62,26 @@ export default class RedirectInterceptor extends Interceptor {
...
       return this.spark.request(options);
     }
+    else if (response.headers
+      && response.body
+      && response.body.errorCode === LOCUS_REDIRECT_ERROR
+      && response.body.location) {
+      options = clone(options);
+      if (response.options && response.options.qs) {
+        const newUrl = response.body.location.split('?');
+        options.uri = newUrl[0]; // params are already present in the qs
```
Comment (saurjai3) [inline packages/node_modules/@ciscospark/spark-core/src/interceptors/redirect.js]: for POST

Comment (mattnorris) [reply packages/node_modules/@ciscospark/spark-core/src/interceptors/redirect.js]: I would actually add these as comments to the file itself.

**Pull request merged**